File: backend/services/client_requirements.py
"""What Papol requires of the installed clients that talk to it.

A floor rather than a switch: the server says which versions it can still
speak to, and the client decides what to do about being below it. Saying it
this way means a new release stops being blocked without anyone editing
anything, and there is no single flag that turns every user off.

Nothing here deletes or rebuilds anything. Being too old to sync is a
different fact from being wrong about what you hold, and they are kept
apart deliberately.
"""

from services.notifications import setting_value

MINIMUM_KEY = "desktop_minimum_version"
RECOMMENDED_KEY = "desktop_recommended_version"
DOWNLOAD_KEY = "desktop_download_url"

DEFAULT_DOWNLOAD_URL = "https://github.com/hflsmax/papol/releases"

# The oldest build that speaks the wire this server speaks.
#
# Not a policy: a fact. A paper is named by the digest of its PDF, and a
# build from before that names one by a UUID this server has never held —
# so there is nothing to say to it, and saying so is kinder than answering
# every request with a puzzle. The constant moves when the wire moves, which
# is what stops a protocol change from shipping without its floor.
#
# The setting can hold clients to something newer than this. It cannot let
# in a build that cannot be talked to.
PROTOCOL_MINIMUM_VERSION = "0.2.0"

SUPPORTED = "supported"
DEPRECATED = "deprecated"
INCOMPATIBLE = "incompatible"

# The client announces itself as "Papol macOS/0.1.2". Anything else is some
# other caller — a browser, curl, a proxy — and is never gated: the floor is
# about the installed application, and the web app ships with the server.
_AGENT_PREFIX = "Papol macOS/"


def parse_version(text: str | None) -> tuple[int, int, int] | None:
    """A dotted release as numbers, or None when it is not one.

    Unparseable is not "old": a version this server cannot read is one it
    has no business judging, so callers treat None as supported.
    """
    if not text:
        return None
    parts = text.strip().split(".")
    if len(parts) != 3:
        return None
    try:
        major, minor, patch = (int(part) for part in parts)
    except ValueError:
        return None
    if major < 0 or minor < 0 or patch < 0:
        return None
    return major, minor, patch


def client_version(user_agent: str | None) -> str | None:
    """The version Papol macOS announced, if this was Papol macOS at all.

    The product name has a space in it, so the agent cannot be read one
    whitespace-separated token at a time: the marker is found in the string
    and the version is whatever follows it up to the next space.
    """
    if not user_agent:
        return None
    start = user_agent.find(_AGENT_PREFIX)
    if start < 0:
        return None
    rest = user_agent[start + len(_AGENT_PREFIX):].split()
    return rest[0] if rest else None


def minimum_version(db) -> str:
    """The floor: whichever is higher of what the wire needs and what was
    asked for. A setting below the protocol's own floor is not a way to let
    an unspeakable build back in; it simply has nothing left to say."""
    asked = setting_value(db, MINIMUM_KEY)
    requested = parse_version(asked)
    if requested is not None and requested > parse_version(PROTOCOL_MINIMUM_VERSION):
        return asked
    return PROTOCOL_MINIMUM_VERSION


def requirements(db) -> dict:
    """What the server asks of its clients, as the app is told it."""
    return {
        "minimum_version": minimum_version(db),
        "recommended_version": setting_value(db, RECOMMENDED_KEY),
        "download_url": setting_value(db, DOWNLOAD_KEY) or DEFAULT_DOWNLOAD_URL,
    }


def verdict(db, user_agent: str | None) -> str:
    """Where this caller stands against the floor.

    Every unknown answers "supported": a caller that is not the application
    and a version string nobody can read are both cases where refusing
    service would be the server inventing a rule it was never given. The
    floor itself is never unknown — the wire has one whether or not anyone
    has asked for a higher.
    """
    version = parse_version(client_version(user_agent))
    if version is None:
        return SUPPORTED
    asked = requirements(db)
    minimum = parse_version(asked["minimum_version"])
    if minimum is not None and version < minimum:
        return INCOMPATIBLE
    recommended = parse_version(asked["recommended_version"])
    if recommended is not None and version < recommended:
        return DEPRECATED
    return SUPPORTED

File: cloudflare/src/clientRequirements.ts
// What Papol requires of a client that talks to it: the schema it was built
// for, and, of the desktop app, a build that speaks to the bucket itself.
//
// One number, `schema_version` in schema/sync_registry.json,
// names the data model and the wire that carries it. A Papol client sends
// the one it was compiled with on every request, and the server compares.
// Equal is supported; anything else is a build that cannot be talked to,
// and is told so with a 426 and somewhere to get the build that can. A
// caller that sends no number is not a Papol client and is not gated —
// unless it names itself as the native app in its User-Agent, which a
// build from before the header did, and predating the header means
// predating this schema.
//
// The desktop app also names its version in that User-Agent, and a build
// older than MINIMUM_DESKTOP_VERSION is refused too: it would send its
// files through routes this Worker no longer has (phase 5, step 15).

import registry from "../../schema/sync_registry.json";

export const SCHEMA_HEADER = "X-Papol-Schema";
export const AGENT_PREFIX = "Papol macOS/";
export const DOWNLOAD_URL = "https://github.com/hflsmax/papol/releases";
export const MINIMUM_DESKTOP_VERSION = "0.5.0";

export const SUPPORTED = "supported";
export const INCOMPATIBLE = "incompatible";

export function schemaVersion(): number {
  return registry.schema_version;
}

// The schema this caller was built for, or null for a caller that says nothing.
export function clientSchema(request: Request): number | null {
  const value = request.headers.get(SCHEMA_HEADER);
  if (value === null) return null;
  const announced = Number(value);
  return Number.isInteger(announced) ? announced : -1; // a header nobody can read is not this build's
}

// The desktop build's version as its User-Agent states it, or null for
// any other caller.
export function desktopVersion(request: Request): string | null {
  const agent = request.headers.get("user-agent") ?? "";
  const start = agent.indexOf(AGENT_PREFIX);
  if (start < 0) return null;
  return agent.slice(start + AGENT_PREFIX.length).split(/\s+/)[0] || "";
}

// Whether a version reads as at least the minimum: numbers compared in
// order, a missing part being zero, and anything unreadable being older.
export function atLeast(version: string, minimum: string): boolean {
  const parts = (v: string) => v.split(".").map((p) => Number.parseInt(p, 10));
  const a = parts(version), b = parts(minimum);
  for (let i = 0; i < Math.max(a.length, b.length); i++) {
    const x = a[i] ?? 0, y = b[i] ?? 0;
    if (!Number.isInteger(x)) return false;
    if (x !== y) return x > y;
  }
  return true;
}

export function verdict(request: Request): string {
  const announced = clientSchema(request);
  const desktop = desktopVersion(request);
  if (desktop !== null && !atLeast(desktop, MINIMUM_DESKTOP_VERSION)) return INCOMPATIBLE;
  if (announced === null) return SUPPORTED;
  return announced === schemaVersion() ? SUPPORTED : INCOMPATIBLE;
}

// What the server asks of its clients, as the app is told it, and where
// the files are: the bucket's own address, or null when this Worker
// serves them itself.
export function requirements(env?: Env) {
  return {
    schema_version: schemaVersion(), download_url: DOWNLOAD_URL, minimum_desktop_version: MINIMUM_DESKTOP_VERSION,
    files_url: env?.FILES_URL ? env.FILES_URL.replace(/\/+$/, "") : null,
  };
}

// The body of a 426: what was refused, and where to go.
export function refusal() {
  return { error: "client_incompatible", ...requirements() };
}
